lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -508,6 +529,18 @@ bool SILCombiner::optimizeStackAllocatedEnum(AllocStackInst *AS) {
         eraseInstFromFunction(*user);
         break;
       case SILInstructionKind::DestroyAddrInst:
+        if (singleInitTakePair) {
+          // It's not possible that the enum has a payload at the destroy_addr,
+          // because it must have already been taken by the take of the
+          // single init-take pair.
+          // We _have_ to remove the destroy_addr, because we also remove all
+          // inject_enum_addrs which might inject a payload-less case before
+          // the destroy_addr.
```

Comment:
Ok, based on this comment, the question for anyone reading the code should be: why aren't we always deleting the destroy_addr then? Reading between the lines, I'm guessing we sometimes have a destroy_addr of the payload enum without a "take". I think there needs to be a similar explanation for why the "else" case is always correct now. After this change, there are subtle implicit dependencies between these three sections of changed code that are just becoming clear after staring at it for a long time.